File: web/providers/twitch.js
// Twitch Helix client. Loads the app access token from ./twitch_token.json
// (written by the CI mint script before deploy). Mirrors providers/twitch.py.

const API = "https://api.twitch.tv/helix";
const WINDOW_DAYS = { day: 1, week: 7, month: 30 };

let _auth = null;

async function getAuth() {
  if (_auth) return _auth;
  const resp = await fetch("./twitch_token.json", { cache: "no-cache" });
  if (!resp.ok) {
    throw new Error(
      `No Twitch token bundled (HTTP ${resp.status}). The CI deploy mints one.`
    );
  }
  _auth = await resp.json();
  return _auth;
}

async function helix(path, params) {
  const { client_id, access_token } = await getAuth();
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params || {})) {
    url.searchParams.set(k, String(v));
  }
  const resp = await fetch(url, {
    headers: {
      "Client-ID": client_id,
      Authorization: `Bearer ${access_token}`,
    },
  });
  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new Error(`Twitch ${resp.status}: ${text || resp.statusText}`);
  }
  return resp.json();
}

async function getUserId(login) {
  const data = await helix("/users", { login });
  return (data && data.data && data.data[0] && data.data[0].id) || null;
}

function helixDate(d) {
  // 2026-04-27T12:00:00Z (no millis, RFC3339-Z)
  return d.toISOString().replace(/\.\d{3}Z$/, "Z");
}

function buildEmbedUrl(slug) {
  const parent = window.location.hostname || "localhost";
  const params = new URLSearchParams({
    clip: slug,
    parent,
    autoplay: "true",
    muted: "false",
  });
  return `https://clips.twitch.tv/embed?${params.toString()}`;
}

function resolveThumbnail(raw) {
  // Twitch returns a template URL with {width} and {height} placeholders.
  if (!raw) return "";
  return raw.replace("{width}", "168").replace("{height}", "94");
}

function toClip(c) {
  return {
    provider: "twitch",
    id: c.id,
    title: c.title || "",
    viewCount: Number(c.view_count || 0),
    durationSec: Number(c.duration || 0),
    thumbnailUrl: resolveThumbnail(c.thumbnail_url),
    embedUrl: buildEmbedUrl(c.id),
    url: c.url || "",
    broadcasterName: c.broadcaster_name || "",
    createdAt: c.created_at || "",
  };
}

export async function getTopClips(login, win, count = 50) {
  const userId = await getUserId(login);
  if (!userId) return [];
  const ended = new Date();
  const days = WINDOW_DAYS[win] ?? 7;
  const started = new Date(ended.getTime() - days * 86400000);
  const data = await helix("/clips", {
    broadcaster_id: userId,
    started_at: helixDate(started),
    ended_at: helixDate(ended),
    first: Math.min(Math.max(count, 1), 100),
  });
  return (data.data || []).map(toClip);
}

File: src/clip_review/providers/twitch.py
"""Twitch Helix provider for top-by-views clips within a time window.

Reference for the JavaScript client running in the browser. Production fetches
happen there; this module mirrors the same logic and is exercised by tests.
"""

from __future__ import annotations

from datetime import datetime
from types import TracebackType
from typing import Any, Self

import httpx

from clip_review.providers import Clip, Window, window_to_range

API_BASE = "https://api.twitch.tv/helix"
TOKEN_URL = "https://id.twitch.tv/oauth2/token"


def _to_clip(raw: dict[str, Any]) -> Clip:
    return Clip(
        provider="twitch",
        id=str(raw["id"]),
        title=str(raw.get("title", "")),
        view_count=int(raw.get("view_count", 0)),
        duration_sec=float(raw.get("duration", 0.0)),
        thumbnail_url=str(raw.get("thumbnail_url", "")),
        embed_url=str(raw.get("embed_url", "")),
        url=str(raw.get("url", "")),
        broadcaster_name=str(raw.get("broadcaster_name", "")),
        created_at=str(raw.get("created_at", "")),
    )


async def mint_app_access_token(
    client_id: str, client_secret: str, *, http: httpx.AsyncClient | None = None
) -> str:
    """Run the OAuth client_credentials flow and return the bearer token.

    Used by the CI cron that refreshes the token written into the static bundle.
    """
    owns = http is None
    client = http or httpx.AsyncClient(timeout=10.0)
    try:
        resp = await client.post(
            TOKEN_URL,
            data={
                "client_id": client_id,
                "client_secret": client_secret,
                "grant_type": "client_credentials",
            },
        )
        resp.raise_for_status()
        token = resp.json().get("access_token")
        if not isinstance(token, str) or not token:
            raise RuntimeError("twitch token response missing access_token")
        return token
    finally:
        if owns:
            await client.aclose()


class TwitchProvider:
    """Async helper around Helix /users and /clips."""

    def __init__(
        self,
        client_id: str,
        app_access_token: str,
        *,
        http: httpx.AsyncClient | None = None,
    ) -> None:
        self._client_id = client_id
        self._token = app_access_token
        self._http = http or httpx.AsyncClient(timeout=10.0)
        self._owns_http = http is None

    async def __aenter__(self) -> Self:
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        if self._owns_http:
            await self._http.aclose()

    @property
    def _headers(self) -> dict[str, str]:
        return {
            "Client-ID": self._client_id,
            "Authorization": f"Bearer {self._token}",
        }

    async def get_user_id(self, login: str) -> str | None:
        resp = await self._http.get(
            f"{API_BASE}/users", params={"login": login}, headers=self._headers
        )
        resp.raise_for_status()
        data = resp.json().get("data") or []
        if not data:
            return None
        return str(data[0]["id"])

    async def get_top_clips(
        self,
        broadcaster_id: str,
        started_at: datetime,
        ended_at: datetime,
        first: int = 50,
    ) -> list[Clip]:
        first_clamped = min(max(first, 1), 100)
        params: dict[str, str | int] = {
            "broadcaster_id": broadcaster_id,
            "started_at": started_at.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "ended_at": ended_at.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "first": first_clamped,
        }
        resp = await self._http.get(f"{API_BASE}/clips", params=params, headers=self._headers)
        resp.raise_for_status()
        raw = resp.json().get("data") or []
        return [_to_clip(c) for c in raw]

    async def top_clips_for_login(self, login: str, window: Window, count: int = 50) -> list[Clip]:
        """Convenience: resolve login → user_id → top clips in one call."""
        user_id = await self.get_user_id(login)
        if user_id is None:
            return []
        started, ended = window_to_range(window)
        return await self.get_top_clips(user_id, started, ended, first=count)
